lib/rubygems/printable_uri.rb

```diff
@@ -0,0 +1,60 @@
+# frozen_string_literal: true
+
+require_relative 'uri_parser'
+
+class Gem::PrintableUri
+  def self.parse_uri(uri)
+    new(uri).parse_uri
+  end
+
+  def initialize(original_uri)
+    @original_uri = original_uri
+  end
+
+  def parse_uri
+    @original_uri = Gem::UriParser.parse_uri(@original_uri)
+    @uri = @original_uri.clone
```

Comment:
I'm not sure whether we want `clone` or `dup` here, but I see that similar code in `Gem::RemoteFetcher` uses `dup... :man_shrugging:.

Reply:
It's refactored to use `dup`.